services/src/contexts/postgres.rs

```diff
@@ -0,0 +1,2597 @@
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new_with_data(
+        config: Config,
+        tls: Tls,
+        dataset_defs_path: PathBuf,
+        provider_defs_path: PathBuf,
+        layer_defs_path: PathBuf,
+        layer_collection_defs_path: PathBuf,
+        exe_ctx_tiling_spec: TilingSpecification,
+        query_ctx_chunk_size: ChunkByteSize,
+    ) -> Result<Self> {
+        let pg_mgr = PostgresConnectionManager::new(config, tls);
+
+        let pool = Pool::builder().build(pg_mgr).await?;
+        let created_schema = Self::create_schema(pool.get().await?).await?;
+
+        let session = if created_schema {
+            let session = SimpleSession::default();
+            Self::create_default_session(pool.get().await?, session.id()).await?;
+            session
+        } else {
+            Self::load_default_session(pool.get().await?).await?
+        };
+
+        let app_ctx = PostgresContext {
+            default_session_id: session.id(),
+            task_manager: Default::default(),
+            thread_pool: create_rayon_thread_pool(0),
+            exe_ctx_tiling_spec,
+            query_ctx_chunk_size,
+            pool,
+            volumes: Default::default(),
+        };
+
+        if created_schema {
+            info!("Populating database with initial data...");
+
+            let ctx = app_ctx.session_context(session);
+
+            let mut db = ctx.db();
+            add_layers_from_directory(&mut db, layer_defs_path).await;
+            add_layer_collections_from_directory(&mut db, layer_collection_defs_path).await;
+
+            add_datasets_from_directory(&mut db, dataset_defs_path).await;
+
+            add_providers_from_directory(&mut db, provider_defs_path, &[]).await;
+        }
+
+        Ok(app_ctx)
+    }
+
+    async fn is_schema_initialized(
+        conn: &PooledConnection<'_, PostgresConnectionManager<Tls>>,
+    ) -> Result<bool> {
+        let stmt = match conn.prepare("SELECT TRUE from geoengine;").await {
+            Ok(stmt) => stmt,
+            Err(e) => {
+                if let Some(code) = e.code() {
+                    if *code == SqlState::UNDEFINED_TABLE {
+                        info!("Initializing schema.");
+                        return Ok(false);
+                    }
+                }
+                return Err(error::Error::TokioPostgres { source: e });
+            }
+        };
+
+        let _row = conn.query(&stmt, &[]).await?;
+
+        Ok(true)
+    }
+
+    #[allow(clippy::too_many_lines)]
+    /// Creates the database schema. Returns true if the schema was created, false if it already existed.
+    pub(crate) async fn create_schema(
+        conn: PooledConnection<'_, PostgresConnectionManager<Tls>>,
+    ) -> Result<bool> {
+        if Self::is_schema_initialized(&conn).await? {
+            return Ok(false);
+        }
+
+        conn.batch_execute(
+            &format!(r#"
+            CREATE TABLE geoengine (
```

Comment:
Is geoengine a meaningful table name? Everything is the geo engine. It is about instance metadata?

Reply:
Yeah I am not sure either. I added it as a replacement for the versions table to check whether the schema is already initialized and as a place to later store stuff like what cargo features were active and whether the schema may be truncated on start. We can give it a different name if you have a better one. Configurations, Settings, ... idk. Maybe it can also be part of the version/migration table we will need for the schema migration.